Enabling DMA on stm32l5 series

ikmdani opened the pull request:
These commits enable DMA in stm32l5 series.
This has been tested with `tests/drivers/dma/loop_transfer`, `tests/drivers/dma/chan_blen_transfer` & `tests/drivers/uart/uart_async_api` on stm32l562e_dk platform, 
`tests/drivers/spi/spi_loopback` & `tests/drivers/uart/uart_async_api` on nucleo_l552ze_q platform.

ikmdani committed:
dts/arm: STM32: Adding DMA support for stm32l5 in dtsi

This commit adds dma1, dma2, dmamux support in dtsi for stm32l5
series. This commit also fixes a compilation warning due to the
usage of upper case in "reg = <0x4000F400 0x400>".

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

ikmdani committed:
tests/drivers: stm32l562e_dk: Configures DMA for chan_blen_transfer test

This commit configures the required DMA parameters to run
tests/drivers/dma/chan_blen_transfer test on stm32l562e_dk platform.
This has been tested and is working as expected.

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

ikmdani committed:
tests/drivers: stm32l562e_dk: Configures DMA to run loop_transfer test.

This commit configures the required DMA parameters to run
tests/drivers/dma/loop_transfer test on stm32l562e_dk platform.
This has been tested and is working as expected.

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

ikmdani committed:
drivers/dma: STM32: Initialising DMA_InitStruct with 0 for stm32l5xx

In stm32l5xx soc, the LL_DMA_InitTypeDef has 2 more fields
(DoubleBufferMode and TargetMemInDoubleBufferMode) that must be
initialised with 0 else the configuration is wrong and gives
wrong values to the LL_DMA_Init function. Due to this the test
tests/drivers/dma/loop_transfer too would fail.

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

ikmdani commented:
@FRASTM

FRASTM commented:
failure does not seem to be related to this PR 

```
INFO    -  49/137 mps2_an521_remote         tests/arch/arm/arm_thread_swap/arch.arm.swap.common.no_optimizations FAILED Build failure (build)
--
```
"

FRASTM commented:
+ adding a test on the nucleo_l552ze is a must:
```
 tests: drivers: spi loopback transfer on stm32l5
    
    This is the board configuration for testing the spi
    dma transfer on the nucleo l552ze board.
    The MISO and MOSI pins of the spi1
    must be connected on the board.
```
[patch0.patch.txt](https://github.com/zephyrproject-rtos/zephyr/files/6731346/patch0.patch.txt)

ikmdani committed:
tests/drivers: nucleo_l552ze_q: Configuration for spi_loopback test

This commit configures spi, dma & dmamux to run spi_loopback test
on nucleo_l552ze_q platform. The tx & rx pin of spi1 should be
shorted before running spi_loopback.

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

FRASTM commented on tests/drivers/spi/spi_loopback/boards/nucleo_l552ze_q.overlay:
```diff
@@ -0,0 +1,34 @@
+	status = "okay";
+};
+
+&rcc {
```
maybe  change of the rcc this is not really needed, could you please test wit and without,

ABOSTM commented on dts/arm/st/l5/stm32l5.dtsi:
```diff
@@ -273,6 +273,43 @@
+			clocks = <&rcc STM32_CLOCK_BUS_AHB1 0x4>;
+			dma-channels = <16>;
+			dma-generators = <4>;
+			dma-requests= <90>;
```
In RM0438 Rev 6, dma there is up to 94 requests

ikmdani replied:
> In RM0438 Rev 6, dma there is up to 94 requests

RM0438 Rev **7** says 90 requests.

FRASTM replied:
Correct, _but_ the count is actually dma-generators + dma-requests = 94, the definition relies on table 96 of the RM0438
(not on list of table 97)

ikmdani replied:
This needs to be corrected in stm32l4r5.dtsi & stm32wb.dtsi. Will raise a PR on this.

erwango replied:
ok, so what is the right value here finally ?

ikmdani replied:
> ok, so what is the right value here finally ?

90 is the right value for L5. Other series are having the right values, no change needed.

ABOSTM replied:
I am still on 94
![image](https://user-images.githubusercontent.com/50730894/124439574-dfc5de00-dd79-11eb-8340-6e9b6ab1cf16.png)

ikmdani replied:
From https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/dma/dmamux_stm32.c#L304:L311, gen_nb gets the dma generators and req_nb gets the dma requests from the DT.

```
const struct dmamux_stm32_config dmamux_stm32_config_##index = {        \
        DMAMUX_CLOCK_INIT(index)                                        \
        .base = DT_INST_REG_ADDR(index),                                \
        .channel_nb = DT_INST_PROP(index, dma_channels),                \
        .gen_nb = DT_INST_PROP(index, dma_generators),                  \
        .req_nb = DT_INST_PROP(index, dma_requests),                    \
        .mux_channels = dmamux_stm32_channels_##index,                  \
};
```

Below https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/dma/dmamux_stm32.c#L43:L53 is the definition of gen_nb and req_nb:

```
/* this is the configuration of the dmamux IP */
struct dmamux_stm32_config {
#if DT_INST_NODE_HAS_PROP(0, clocks)
        struct stm32_pclken pclken;
#endif
        uint32_t base;
        uint8_t channel_nb;     /* total nb of channels */
        uint8_t gen_nb; /* total nb of Request generator */
        uint8_t req_nb; /* total nb of Peripheral Request inputs */
        const struct dmamux_stm32_channel *mux_channels;
};
```

Referring RM0438 Rev7 stm32L562 reference manual,
![image](https://user-images.githubusercontent.com/78502494/124506502-68c82e00-dde9-11eb-9005-8aa065eda9bb.png)

In the below code, the max request Id is calculated by adding req_nb (90 in case of L5) + gen_nb (4 in case of L5).

```

int dmamux_stm32_configure(const struct device *dev, uint32_t id,
                                struct dma_config *config)
{
        /* device is the dmamux, id is the dmamux channel from 0 */
        const struct dmamux_stm32_config *dev_config = dev->config;

        /*
         * request line ID for this mux channel is stored
         * in the dma_slot parameter
         */
        int request_id = config->dma_slot;

        if (request_id > dev_config->req_nb + dev_config->gen_nb) {
                LOG_ERR("request ID %d is not valid.", request_id);
                return -EINVAL;
        }
```

Pls let me know your thoughts on this.

ABOSTM replied:
Ok I agree. Thanks for the explanation.

ABOSTM commented on tests/drivers/spi/spi_loopback/boards/nucleo_l552ze_q.overlay:
```diff
@@ -0,0 +1,26 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&spi1 {
```
Why adding spi1 in the overlay of the test ?
Why not adding spi1 in board Devicetree ?

FRASTM replied:
so that the dts is changed just for _that_ test,

erwango replied:
Please move spi1 definition (pins and status) to board description.
(dma part should stay here)

erwango replied:
Request still stands

ABOSTM commented on tests/drivers/uart/uart_async_api/boards/nucleo_l552ze_q.overlay:
```diff
@@ -0,0 +1,26 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&usart3 {
```
Why adding usart3 in the overlay of the test ?
Why not adding usart3 in board Devicetree ?

FRASTM replied:
Change is limited to _that_ test only

erwango replied:
Since there is quite little devices specified in board description, it would indeed make sense to take advantage of this to add one spi device and one uart device to the boards (specially if these are arduino ones)

ikmdani committed:
tests/drivers: nucleo_l552ze_q: Configures DMA for chan_blen_transfer test

This commit configures the DMA to run
tests/drivers/dma/chan_blen_transfer test on nucleo_l552ze_q platform.

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

ikmdani committed:
tests/drivers: nucleo_l552ze_q: Configures DMA to run loop_transfer test

This commit configures DMA to run tests/drivers/dma/loop_transfer
test on nucleo_l552ze_q platform.

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

erwango commented on tests/drivers/spi/spi_loopback/boards/nucleo_l552ze_q.conf:
```diff
@@ -0,0 +1,8 @@
+CONFIG_SPI_LOOPBACK_DRV_NAME="SPI_1"
+
+CONFIG_SPI_STM32_DMA=n
```
```suggestion
CONFIG_SPI_STM32_DMA=y
```

erwango commented on tests/drivers/uart/uart_async_api/boards/nucleo_l552ze_q.overlay:
```diff
@@ -0,0 +1,26 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&usart3 {
+	pinctrl-0 = <&usart3_tx_pd8 &usart3_rx_pd9>;
+	current-speed = <115200>;
+	dmas = <&dmamux1 5 30 0x440 0>,
+		<&dmamux1 4 29 0x480 0>;
+	dma-names = "tx", "rx";
+	status = "okay";
+};
```
Please move node definition to board and keep only the dma bits here.

erwango commented on tests/drivers/uart/uart_async_api/boards/stm32l562e_dk.overlay:
```diff
@@ -0,0 +1,27 @@
+ */
+
+
+&usart3 {
```
ditto

ikmdani committed:
tests/drivers: stm32l562e_dk: uart_async_api test

This commit configures usart3 to use dma in uart_async_api test
on stm32l562e_dk platform. Short pin2(PC10) & pin 3 (PC11) of usart3
in CN4 connector on stm32l562e_dk platform.

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

ikmdani committed:
tests/drivers: nucleo_l552ze_q: uart_async_api test

This commit configures usart3 to use dma in uart_async_api test
on nucleo_l552ze_q platform.

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

ABOSTM commented on boards/arm/nucleo_l552ze_q/nucleo_l552ze_q.dts:
```diff
@@ -32,3 +32,7 @@
 };
+
+&usart3 {
+	pinctrl-0 = <&usart3_tx_pd8 &usart3_rx_pd9>;
```
Is there a reason to dissociate `pinctrl-0` defintion from `status = "okay"` ?
Why not moving `pinctrl-0` to file **nucleo_l552ze_q-common.dtsi** ?

ABOSTM commented on boards/arm/stm32l562e_dk/stm32l562e_dk.dts:
```diff
@@ -35,6 +35,10 @@
 };
 
+&usart3 {
+	pinctrl-0 = <&usart3_tx_pc10 &usart3_rx_pc11>;
```
Is there a reason to dissociate pinctrl-0 defintion from `status = "okay" `?
Why not moving `pinctrl-0` to file **stm32l562e_dk_common.dtsi**  ?

ikmdani committed:
boards/arm: nucleo_l552ze_q: Updating the documentation

This commit updates the documentation on DMA for nucleo_l552ze_q
platform.

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

ikmdani committed:
boards/arm: nucleo_l552ze_q: Enabling usart3

This commit enables usart3 channel on nucleo_l552ze_q platform.
This would also be used to test usart in dma mode with
tests/drivers/uart/uart_async_api.

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

ikmdani committed:
boards/arm: nucleo_l552ze_q: Enabling spi1

This commit enables spi1 in dtsi and updates the
documentation for nucleo_l552ze_q platform.

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

ikmdani committed:
boards/arm: stm32l562e_dk: Updating the documentation

This commit updates the documentation on DMA for stm32l562e_dk
platform.

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

ikmdani committed:
boards/arm: stm32l562e_dk: Enabling usart3

This commit enables usart3 channel on stm32l562e_dk platform.
This would also be used to test usart in dma mode with
tests/drivers/uart/uart_async_api.

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>
